sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuCartesianProductExec.scala

```diff
@@ -0,0 +1,264 @@
+        JCudfSerialization.writeToStream(columns, out, 0, numRows)
+        columns.safeClose()
+      }
+    }
+  }
+
+  private def readObject(in: ObjectInputStream): Unit = {
+    GpuSemaphore.acquireIfNecessary(TaskContext.get())
+    withResource (new NvtxRange("DeserializeBatch", NvtxColor.PURPLE)) { _ =>
+      withResource(JCudfSerialization.readTableFrom(in)) { tableInfo =>
+        val tmp = tableInfo.getTable
+        if (tmp == null) {
+          throw new IllegalStateException("Empty Batch???")
+        }
+        this.internalBatch = GpuColumnVector.from(tmp)
+      }
+    }
+  }
+
+  override def close(): Unit = {
+    if (internalBatch != null) {
+      internalBatch.close()
+    }
+  }
+}
+
+class GpuCartesianPartition(
+    idx: Int,
+    @transient private val rdd1: RDD[_],
+    @transient private val rdd2: RDD[_],
+    s1Index: Int,
+    s2Index: Int
+) extends Partition {
+  var s1: Partition = rdd1.partitions(s1Index)
+  var s2: Partition = rdd2.partitions(s2Index)
+  override val index: Int = idx
+
+  @throws(classOf[IOException])
+  private def writeObject(oos: ObjectOutputStream): Unit = Utils.tryOrIOException {
+    // Update the reference to parent split at the time of task serialization
+    s1 = rdd1.partitions(s1Index)
+    s2 = rdd2.partitions(s2Index)
+    oos.defaultWriteObject()
+  }
+}
+
+class GpuCartesianRDD(
+    sc: SparkContext,
+    boundCondition: Option[GpuExpression],
+    joinTime: SQLMetric,
+    joinOutputRows: SQLMetric,
+    numOutputRows: SQLMetric,
+    numOutputBatches: SQLMetric,
+    filterTime: SQLMetric,
+    totalTime: SQLMetric,
+    var rdd1 : RDD[GpuSerializableBatch],
+    var rdd2 : RDD[GpuSerializableBatch])
+    extends RDD[ColumnarBatch](sc, Nil)
+        with Serializable with Arm {
+
+  private val numPartitionsInRdd2 = rdd2.partitions.length
+
+  override def getPartitions: Array[Partition] = {
+    // create the cross product split
+    val array = new Array[Partition](rdd1.partitions.length * rdd2.partitions.length)
+    for (s1 <- rdd1.partitions; s2 <- rdd2.partitions) {
+      val idx = s1.index * numPartitionsInRdd2 + s2.index
+      array(idx) = new GpuCartesianPartition(idx, rdd1, rdd2, s1.index, s2.index)
+    }
+    array
+  }
+
+  override def getPreferredLocations(split: Partition): Seq[String] = {
+    val currSplit = split.asInstanceOf[GpuCartesianPartition]
+    (rdd1.preferredLocations(currSplit.s1) ++ rdd2.preferredLocations(currSplit.s2)).distinct
+  }
+
+  override def compute(split: Partition, context: TaskContext):
+  Iterator[ColumnarBatch] = {
+    val currSplit = split.asInstanceOf[GpuCartesianPartition]
+    rdd1.iterator(currSplit.s1, context).flatMap { lhs =>
+      val table = withResource(lhs) { lhs =>
+        GpuColumnVector.from(lhs.getBatch)
+      }
+      // Ideally instead of looping through and recomputing rdd2 for
+      // each batch in rdd1 we would instead cache rdd2 in a way that
```

Comment:
I think this is relatively straightforward to do if we always configure the memory store hierarchy (i.e.: dev->host->disk) rather than only conditionally when using the UCX shuffle.  We'd probably want to make a contiguous block of memory for each batch coming out of `rdd2.iterator`, but once we have that we could call `RapidsDeviceMemoryStore.addTable` to add the table to the store.  Then on the second and subsequent iterations of the `rdd1.iterator` loop, we'd acquire the buffer from `RapidsBufferCatalog` and unlock the buffer after the join completes.  On the last iteration we'd free the buffer since it's no longer needed.  The spill store will automatically handle spilling the partitions to host memory and disk as necessary.